### DISCUSSION TITLE
Building in I2S support

### DISCUSSION THREAD

**Discussion by mrburns42** [micropython/micropython#9151] [Libraries & Drivers] [answered]
I am trying to build a ESP32 binary with I2S support.   According to the micropython website, it is available as a "technical preview"   .  However, when I build from the latest code,  there is no I2S module in the build,    I searched and I cannot find a definition of what a "technical preview" is , how to download it, or how to build that version.    
How do I enable I2S in the build?    Is this covered by some overall configurator?    

As a noobie,  I have no clue where to look and I am not really finding a startup guide that would walk me through this kind of issue. 
Thanks.

**Comment by Lobo-T** [comment]
I think it gets built by default for the ESP versions that support it?

Seems to be in my build atleast:
```
>>> help(machine)
object <module 'umachine'> is of type module
  __name__ -- umachine
  mem8 -- <8-bit memory>
  mem16 -- <16-bit memory>
  mem32 -- <32-bit memory>
  freq -- <function>
  reset -- <function>
  soft_reset -- <function>
  unique_id -- <function>
  sleep -- <function>
  lightsleep -- <function>
  deepsleep -- <function>
  idle -- <function>
  disable_irq -- <function>
  enable_irq -- <function>
  bitstream -- <function>
  time_pulse_us -- <function>
  Timer -- <class 'Timer'>
  WDT -- <class 'WDT'>
  SDCard -- <class 'SDCard'>
  SLEEP -- 2
  DEEPSLEEP -- 4
  Pin -- <class 'Pin'>
  Signal -- <class 'Signal'>
  TouchPad -- <class 'TouchPad'>
  ADC -- <class 'ADC'>
  ADCBlock -- <class 'ADCBlock'>
  DAC -- <class 'DAC'>
  I2C -- <class 'I2C'>
  SoftI2C -- <class 'SoftI2C'>
  I2S -- <class 'I2S'>
  PWM -- <class 'PWM'>
  RTC -- <class 'RTC'>
  SPI -- <class 'SPI'>
  SoftSPI -- <class 'SoftSPI'>
  UART -- <class 'UART'>
  reset_cause -- <function>
  HARD_RESET -- 2
  PWRON_RESET -- 1
  WDT_RESET -- 3
  DEEPSLEEP_RESET -- 4
  SOFT_RESET -- 5
  wake_reason -- <function>
  PIN_WAKE -- 2
  EXT0_WAKE -- 2
  EXT1_WAKE -- 3
  TIMER_WAKE -- 4
  TOUCHPAD_WAKE -- 5
  ULP_WAKE -- 6
>>> 

```

**Reply by mrburns42** [reply]
Thanks.   I am using the Stamp Pico.    I2S is supported and I got it to work from C++.    However,  when I tried pre-built binaries, the only one that would not boot loop was the generic.    So, when I built from scratch,  I also built the generic board.  Perhaps that is my problem and I need to build for the Pico.     I will try the build again.

**Comment by jimmo** [comment, marked as answer]
It is enabled on ESP32 for all builds and boards except the ESP32C3. Is that what you have?

I do not know why it isn't enabled on the C3... I know the C3 does have an I2S peripheral. It's possible that support wasn't added until a later IDF version, or possibly just an oversight.

**Reply by mrburns42** [reply]
Thanks.   I built for the Generic.   Perhaps it is not enabled there.   I will try again.   With the C3,   I tried that board under C++ and I2S does not work there either.    I believe the C3 only supports I2S as an output device.   There is some silicon problem with I2S inputs on the C3.  That is probably why it is disabled.

**Reply by jimmo** [reply]
It is enabled for everything except the C3-based boards.

If you build the GENERIC firmware (i.e. for regular ESP32), you should have it. See esp32/mpconfigport.h where it has

```
#ifndef MICROPY_PY_MACHINE_I2S
#define MICROPY_PY_MACHINE_I2S              (1)
#endif
```

The only boards that override this are the C3 ones.

Can you do this at the REPL:

```
>>> import machine
>>> machine.I2S
<class 'I2S'>
```

**Reply by mrburns42** [reply]
Below is the output.    It does not seem to find the I2S module.    Based on the date, it seems to be running the version that I built last week. 
Probably operator error, as I am brand-new to micropython and not that hot in python itself.    


MicroPython v1.19.1-316-g8139cbcf6 on 2022-08-26; ESP32 module with ESP32

Type "help()" for more information.
>>> import machine
>>> machine.i2S
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'i2S'
>>> 


>>> help ('modules')
__main__          framebuf          uasyncio/stream   uplatform
_boot             gc                ubinascii         urandom
_onewire          inisetup          ubluetooth        ure
_thread           math              ucollections      uselect
_uasyncio         micropython       ucryptolib        usocket
_webrepl          neopixel          uctypes           ussl
apa106            network           uerrno            ustruct
btree             ntptime           uhashlib          usys
builtins          onewire           uheapq            utime
cmath             uarray            uio               utimeq
dht               uasyncio/__init__ ujson             uwebsocket
ds18x20           uasyncio/core     umachine          uzlib
esp               uasyncio/event    uos               webrepl
esp32             uasyncio/funcs    upip              webrepl_setup
flashbdev         uasyncio/lock     upip_utarfile
Plus any modules on the filesystem

**Reply by jimmo** [reply]
> import machine
> machine.i2S
> Traceback (most recent call last):
> File "", line 1, in
> AttributeError: 'module' object has no attribute 'i2S'

It's `I2S` not `i2S`.

**Reply by mrburns42** [reply]
Yep, you are correct.  When I spell it with the proper capitalization, then it is found.   My error is in spelling and assuming that if I list the modules, then it would be in that list.   


>>> machine.I2S
<class 'I2S'>
>>>

**Reply by jimmo** [reply]
Listing the modules won't see it -- `I2S` is a _class_ in the `machine` _module_. You can see `machine` in your list above from `help('modules')`.

See @Lobo-T 's reply above where they did `help(machine)` to list the contents of the machine module (which included I2S).

**Reply by mrburns42** [reply]
Thank you for all the help.   Root cause was just operator error on my part.    As I said,  I am brand new to micropython and weak in python.   I started with a tutorial and forum posts that predated the inclusion of I2S into the repo.  By mixing and jumbling their contents, I got my mind set that it must show up as a module.  Then when I did get the latest build installed, I assumed it was missing when it as there all the time.  Thanks again.

**Comment by walklinewang** [comment]
Is there any possible the config file disabled `I2S` class for some reason? ![link](https://github.com/micropython/micropython/blob/master/ports/esp32/boards/GENERIC_C3/mpconfigboard.h#L8)

**Reply by jimmo** [reply]
See suggestion above as to why this might be disabled: https://github.com/orgs/micropython/discussions/9151#discussioncomment-3509255
